docs/debugging.rst

```diff
@@ -0,0 +1,249 @@
+        kern()
+.. note::
+
+    For the **OpenGL and CUDA backend**, the printed result won't shows up until ``ti.sync()``:
+
+    .. code-block:: python
+
+        import taichi as ti
+        ti.init(arch=ti.cuda)
+
+        @ti.kernel
+        def kern():
+            print('inside kernel')
+
+        print('before kernel')
+        kern()
+        print('after kernel')
+        ti.sync()
+        print('after sync')
+
+    obtains:
+
+    .. code-block:: none
+
+        before kernel
+        after kernel
+        inside kernel
+        after
+
+    Also note that host access or program end will also implicitly invoke for ``ti.sync()``.
+
+
+Compile-time ``ti.static_print``
+--------------------------------
+
+Sometimes it's also useful to print Python-scope objects and constants like datatype or SNode in Taichi-scope.
+So, similar to ``ti.static`` we provide ``ti.static_print`` to print compile-time constants.
+It behaves the same as Python-scope ``print``, just being embedded into Taichi kernel.
+
+.. code-block:: python
+
+    x = ti.var(ti.f32, (2, 3))
+    y = 1
+
+    @ti.kernel
+    def inside_taichi_scope():
+        ti.static_print(y)
+        # => 1
+        ti.static_print(x.shape)
+        # => (2, 3)
+        ti.static_print(x.data_type())
+        # => DataType.float32
+        for i in range(4):
+                ti.static_print(i.data_type())
+                # => DataType.int32
+                # will only print once
+
+Unlike ``print``, ``ti.static_print`` will only print the expression once at compile-time, and
+therefore has no runtime cost.
+
+
+Runtime ``assert`` in kernel
+----------------------------
+
+We may use ``assert`` statement in Taichi-scope. When assertion condition failed, a
+``RuntimeError`` will be raised to indicate error.
+
+To make ``assert`` work, first make sure you are using the **CPU backend**.
+For performance reason, ``assert`` only works when ``debug`` mode is on, For example:
+
+.. code-block:: python
+
+    ti.init(arch=ti.cpu, debug=True)
+
+    x = ti.var(ti.f32, 128)
+
+    @ti.kernel
+    def do_sqrt_all():
+        for i in x:
+            assert x[i] >= 0
+            x[i] = ti.sqrt(x)
+
+
+When your debugging work is done, simply set ``debug=False``. Now ``assert`` will be ignored
+therefore no runtime overhead, making your program easy to debug without losing performance.
+
+
+Compile-time ``ti.static_assert``
+---------------------------------
+
+Like ``ti.static_print``, we also provide a static version of ``assert``:
+``ti.static_assert``, it can be useful to make assertion on data type / dimention / shape.
+It works no matter whether ``debug=True`` is specified. When assertion fails, it will
+raise ``AssertionError`` as a Python-scope ``assert`` does.
+
+For example:
+
+.. code-block:: python
+
+    @ti.func
+    def is_odd(x: ti.template()):
+        ti.static_assert(x.data_type() == ti.i32, "is_odd() is only supported for i32")
```

Comment:
This is a bit misleading. This function seems to take a local scalar, but for now, Taichi cannot infer data types of local variables at compile-time.

Reply:
Oh sorry! Then how about:
```suggestion
    def copy(dst: ti.template(), src: ti.template()):
        ti.static_assert(dst.shape == src.shape, "copy() needs src and dst tensors to be same shape")
        for I in ti.grouped(src):
            dst[I] = src[I]
```